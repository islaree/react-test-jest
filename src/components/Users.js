import { useEffect, useState } from "react";

const getUsers = async () => {
  const res = await fetch("https://jsonplaceholder.typicode.com/users");
  const data = await res.json();
  return data;
};

export const Users = () => {
  const [users, setUsers] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    const fetchUsers = async () => {
      try {
        const data = await getUsers();
        setUsers(data);
      } catch (err) {
        setError("Failed to fetch users");
      } finally {
        setLoading(false);
      }
    };

    fetchUsers();
  }, []);

  if (loading) return <p>Loading...</p>;
  if (error) return <p>{error}</p>;

  return (
    <ul>
      {users.map((user) => (
        <li key={user.id}>{user.name}</li>
      ))}
    </ul>
  );
};
