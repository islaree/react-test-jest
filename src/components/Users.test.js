import { render, screen, waitFor } from "@testing-library/react";
import { Users } from "./Users";

beforeEach(() => {
  global.fetch = jest.fn(
    () =>
      new Promise((resolve) =>
        setTimeout(
          () =>
            resolve({
              json: () =>
                Promise.resolve([
                  { id: 1, name: "John Doe" },
                  { id: 2, name: "Jane Doe" },
                ]),
            }),
          100
        )
      )
  );
});

afterEach(() => {
  jest.resetAllMocks();
});

test("APIから取得したユーザーが正しくレンダリングされる", async () => {
  render(<Users />);

  expect(screen.getByText("Loading...")).toBeInTheDocument();

  await waitFor(() => {
    expect(screen.queryByText("Loading...")).not.toBeInTheDocument();
    expect(screen.getByText("John Doe")).toBeInTheDocument();
    expect(screen.getByText("Jane Doe")).toBeInTheDocument();
  });

  expect(global.fetch).toHaveBeenCalledTimes(1);
  expect(global.fetch).toHaveBeenCalledWith(
    "https://jsonplaceholder.typicode.com/users"
  );
});

test("APIが失敗した時のエラーメッセージが表示される", async () => {
  global.fetch = jest.fn(() =>
    Promise.reject(new Error("Failed to fetch users"))
  );

  render(<Users />);

  expect(await screen.findByText("Failed to fetch users")).toBeInTheDocument();
});

test("Usersコンポーネントのスナップショットが一致するか確認する", () => {
  const { asFragment } = render(<Users />);
  expect(asFragment()).toMatchSnapshot();
});
