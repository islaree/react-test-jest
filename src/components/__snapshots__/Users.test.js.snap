// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Usersコンポーネントのスナップショットが一致するか確認する 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;
